logic for singleTargetTower.java

CVaranese opened the pull request:
Assuming some of the logic for the location class

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/singleTargetTower.java:
```diff
@@ -0,0 +1,35 @@
+
+import java.util.ArrayList;
+
+public class singleTargetTower implements Tower {
```
We should also have a Priority with a getter and setter (see lombok's `@getter` and `@setter`)

thewells1024 commented:
ticket "https://github.com/thewells1024/TowerOffense/issues/59"

Remember to add the issue to the commit message

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/singleTargetTower.java:
```diff
@@ -0,0 +1,35 @@
+      return location;
+   }
+
+   public void selectTarget(ArrayList<Unit> units){
```
Also, this should be a List, not an ArrayList

primitiveDataType commented on app/src/main/java/cpe307/team6/toweroffense/singleTargetTower.java:
```diff
@@ -0,0 +1,35 @@
+
+   public void selectTarget(ArrayList<Unit> units){
+      for(Unit u : units){
+         if(u.getLocation().getDistance(this.location) < attackDistance){
```
For the AOETower, I was planning to put a getDistanceFrom function inside the AOETower class, so that the tower calculates it's distance from units.

Maybe we should consider adding this to the Tower interface when we talk about it.

thewells1024 replied:
It shouldn't be in the tower, I'm ok with putting it in the Location class, I'm just thinking about whether we'll need that logic for different things, I don't think so, but if we find that we start using that logic all over the place, then we can talk about moving it to a static utility class

CVaranese commented:
Changed most of the things you said. I wasn't sure what we're doing with the priority class, since I thought that would be in the Tower interface, so I just implemented as best I could. I'm also not sure if I used Lombok correctly.

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,22 @@
+
+import java.util.List;
+
+@AllArgsConstructor(access = AccessLevel.PUBLIC)
```
Not a huge thing, but PUBLIC is the default access level.

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,22 @@
+   private final static int damage;
+
+   /* returns the first unit in the list that is within the tower's range */
+   public Unit selectTargetByPriority(List<Unit> units){
```
This should return `List<Unit>`

thewells1024 replied:
This won't be sorted by priority, you need to implement the decision logic to select what unit to target. I'd recommend a switch with a separate method for each priority (ie `selectFirstTarget`, `selectLastTarget`, `selectClosestTarget` and `selectHealthiestTarget`)

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,22 @@
+import java.util.List;
+
+@AllArgsConstructor(access = AccessLevel.PUBLIC)
+public class SingleTargetTower implements Tower {
```
This tower will need to store the path that the units use for unit selection for the first and last priority. Also, it should be final, because the path shouldn't change after it's set in the constructor

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,22 @@
+
+   /* returns the first unit in the list that is within the tower's range */
+   public Unit selectTargetByPriority(List<Unit> units){
+	   /* assuming this list is organized by priority already */
```
This won't be sorted by priority, you need to implement the decision logic to select what target to

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,22 @@
+/* design taken from the not-yet-pulled tower interface */
```
This comment should be removed

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,22 @@
+   private final static int attackDistance;
+   private final static int damage;
+
+   /* returns the first unit in the list that is within the tower's range */
```
I'm iffy about whether this comment is really necessary

CVaranese commented:
Still need to add unit tests

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,88 @@
+
+   public List<Unit> selectFirstTarget(List<Unit> units){
+      List<Unit> returnedUnit = new ArrayList<Unit>();
+      for(Unit unit: units) {
```
you may want to define a private helper method `double distanceTo(Unit unit)`, seeing as you'll need it in at least 5 places in this class. However figuring out which one is first, is a little tricky since you only have the path that they;re following, and that requires estimating the distance along the path, I would recommend rounding the location of the unit down to whole numbers, which should coincide with a Location in the path variable, then counting the number of locations after that, and then taking the decimal part of the x and y properties of the unit's location (only one of them should have a non-integer value) and adding `1 - (x % 1) - (y%1)` to that calculation. this should also help with the last priority. You may want to make a private helper method to calculate this.

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,88 @@
+   }
+
+   public List<Unit> selectFirstTarget(List<Unit> units){
+      List<Unit> returnedUnit = new ArrayList<Unit>();
```
This shouldn't be necessary

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,88 @@
+      }
+   }
+
+   public List<Unit> selectLastTarget(List<Unit> units){
```
all arguments should be final

thewells1024 replied:
See my comment about calculating the distance to the base

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,88 @@
+
+   public List<Unit> selectLastTarget(List<Unit> units){
+      List<Unit> returnedUnit = new ArrayList<Unit>();
+      for(int i; i >= 0; i--){
```
You need to check all units because the list isn't guaranteed to be sorted in any particular order. I would simply use a normal for-each loop

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,88 @@
+      }
+   }
+
+   public List<Unit> selectClosestTarget(List<Unit> units){
```
All arguments should be final

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,88 @@
+      double minDistance = Integer.MAX_VALUE;
+      for(Unit unit : units){
+         double distance = unit.getLocation().getDistance(this.location);
+         if(distance < attackDistance && distance < minDistance) {
```
I would just keep it in a variable `Unit currentClosest = null` and then only check that the unit you're looking at is closest if currentClosest isn't null

CVaranese replied:
if this code works, why are you asking me to change it? im better off spending time working on something else than changing something because you dont like the way it looks

thewells1024 replied:
because it's not that big of a change, but it avoids unnecessary branches that make the code more complex and harder to maintain.
```java
Unit closestUnit = null;
double minDistance = Integer.MAX_VALUE;
for (Unit unit : units) {
   double distance = unit.getLocation().getDistance(location);
   if (distance <= attackDistance && distance < minDistance) {
      closestUnit = unit;
      minDistance = distance;
   }
   return closestUnit != null ? Collections.singletonList(closestUnit) : new ArrayList<>();
}
```
is much easier to read, understand, maintain, and test.

Also, as it currently stands, it won't compile because it's expecting a List of Units, and you're passing back a List of Lists of Units

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,88 @@
+   }
+
+   public List<Unit> selectHealthiestTarget(List<Unit> units){
+      List<Unit> returnedUnit = new ArrayList<Unit>();
```
See my comment about closest target, the same logic applies

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,88 @@
+      return returnedUnit;
+   }
+
+   public List<Unit> selectHealthiestTarget(List<Unit> units){
```
All Arguments should be final

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,88 @@
+import java.util.List;
+import java.util.ArrayList;
```
After all the changes I believe this shouldn't be necessary

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,112 @@
+      for(Unit unit : units){
+         double distance = unit.getLocation().getDistance(this.location);
+         int curHealth = unit.getHealth()
+         if(distance < attackDistance && curHealth > maxHealth) {
```
this should be `distance <= attackDistance`

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,112 @@
+         
+   }
+
+   private double getDistance(Location loc1, Location loc2){
```
This can be removed, look at PR https://github.com/thewells1024/TowerOffense/pull/75 `getDistance` is a method of location

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,112 @@
+      for(Unit unit: units) {
+         if(distanceToBase(unit.getLocation())
+            > distanceToBase(returnedUnit.getLocation())
+            && getDistance(returnedUnit.getLocation(), unit.getLocation()
```
this should be `distance <= attackDistance`

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,112 @@
+      for(Unit unit: units) {
+         if(distanceToBase(unit.getLocation())
+            > distanceToBase(returnedUnit.getLocation())
+            && getDistance(returnedUnit.getLocation(), unit.getLocation()
```
this should be `distance <= attackDistance`

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,112 @@
+         case FIRST:
+            return selectFirstTarget(units);
+         case LAST:
+            return selectLastTarget(units);
```
You need a closing brace

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,112 @@
+      }
+      Location pathLoc = new Location((int)loc.getx(), (int)loc.gety()); 
+      int unitIndex = path.indexOf(pathLoc);
+      return Math.abs(index - unitIndex);
```
this won't return the correct value, you need to take into account the fractional part of the location given

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,112 @@
+            returnedUnit = unit;
+         }
+      }
+      return singletonList(unit)
```
This won't compile, it needs to be `return closestUnit != null ? Collections.singletonList(closestUnit) : new ArrayList<>();` or if you static import `Collections.singletonList` then `return closestUnit != null ? singletonList(closestUnit) : new ArrayList<>();`

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,112 @@
+            
+         }
+      }
+      return singletonList(unit);
```
This won't compile, it needs to be `return closestUnit != null ? Collections.singletonList(closestUnit) : new ArrayList<>();` or if you static import `Collections.singletonList` then `return closestUnit != null ? singletonList(closestUnit) : new ArrayList<>();`

CVaranese replied:
do I need to import ArrayList then?

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,112 @@
+   public List<Unit> selectFirstTarget(final List<Unit> units){
+      Unit returnedUnit;
+      for(Unit unit: units) {
+         if(distanceToBase(unit.getLocation())
```
I would switch the order of the booleans on this if statement, there's no point calculating the distance to the base of things outside the attack range

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,112 @@
+            minDistance = distance;
+         }
+      }
+      return singletonList(returnedUnit);
```
This won't compile, it needs to be `return closestUnit != null ? Collections.singletonList(closestUnit) : new ArrayList<>();` or if you static import `Collections.singletonList` then `return closestUnit != null ? singletonList(closestUnit) : new ArrayList<>();`

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,112 @@
+   }
+
+   public List<Unit> selectClosestTarget(final List<Unit> units){
+      List<Unit> returnedUnit = new ArrayList<Unit>();
```
this should be changed to a Unit rather than a List of Units

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,112 @@
+            maxHealth = health;
+         }
+      }
+      return singletonList(returnedUnit);
```
This won't compile, it needs to be `return closestUnit != null ? Collections.singletonList(closestUnit) : new ArrayList<>();` or if you static import `Collections.singletonList` then `return closestUnit != null ? singletonList(closestUnit) : new ArrayList<>();`

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,112 @@
+   }
+
+   public List<Unit> selectHealthiestTarget(final List<Unit> units){
+      List<Unit> returnedUnit = new ArrayList<Unit>();
```
this should be changed to a Unit rather than a List of Units

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,99 @@
+      else{
+         index = path.size() - 1;
+      }
+      Location pathLoc = new Location(Math.round(loc.getx()),
```
This still won't work. You just have to do what you did before, and calculate the fractional part separately

CVaranese replied:
I don't understand this part. If I round it, shouldn't it be on the path?

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,99 @@
+
+   private double distanceToBase(Location loc){
+      int index;
+      if (player = 1){
```
I need to talk to the team about this, but I think we should give the path to the tower so that units start at 0 and end at the last index. But for now, don't change it

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,99 @@
+                                      Math.round(loc.gety())); 
+      int unitIndex = path.indexOf(pathLoc);
+      return Math.abs(index - unitIndex);
+         
```
delete this empty line

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,99 @@
+         }
+      }
+      return returnedUnit != null ? Collections.singletonList(unit) :
+                                    new ArrayList<>();
```
don't put it all the way in, just do 3 spaces

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,99 @@
+         }
+      }
+      return returnedUnit != null ? Collections.singletonList(unit) :
+                                    new ArrayList<>();
```
don't put it all the way in, just do 3 spaces

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,99 @@
+         }
+      }
+      return returnedUnit != null ? Collections.singletonList(unit) :
+                                    new ArrayList<>();
```
don't put it all the way in, just do 3 spaces

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,99 @@
+         }
+      }
+      return returnedUnit != null ? Collections.singletonList(unit) :
+                                    new ArrayList<>();
```
don't put it all the way in, just do 3 spaces

CVaranese commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,99 @@
+      for(Unit unit: units) {
+         if(returnedUnit.getLocation().getDistance(unit.getLocation())
+            <= attackDistance && distanceToBase(unit.getLocation())
+            > distanceToBase(returnedUnit.getLocation())){
```
nope, i just initially thought the statement wouldn't fit on one line for the others, but the code has changed a little bit.

CVaranese replied:
Oh, its actually because i use it twice for those and only once for the other two methods

thewells1024 replied:
I mean you recompute it every time you go through the loop, so it would probably be a good idea to save it

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,95 @@
+      for(Unit unit : units){
+         double distance = unit.getLocation().getDistance(this.location);
+         if(distance <= attackDistance && distance < minDistance) {
+            returnedUnit = unit
```
you need a semi-colon

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,95 @@
+         double distance = unit.getLocation().getDistance(this.location);
+         if(distance <= attackDistance && distance < minDistance) {
+            returnedUnit = unit
+            minDistance = distance
```
you need a semi-colon

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,95 @@
+      Unit returnedUnit;
+      double distance;
+      for(Unit unit: units) {
+         distance = distanceToBase(returnedUnit.getLocation());
```
You're still computing the value every time, it should be something like
```java
double minDistance = Integer.MAX_VALUE;
for(Unit unit : units){
   double distance = distanceToBase(unit.getLocation());
   if(distance <= attackDistance && distance < minDistance) {
      returnedUnit = unit;
      minDistance = distance;
   }
}
```

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,95 @@
+   public List<Unit> selectLastTarget(final List<Unit> units){
+      Unit returnedUnit;
+      double distance;
+      for(Unit unit: units) {
```
You're still computing the value every time, it should be something like
```java
double maxDistance = 0;
for(Unit unit : units){
   double distance = distanceToBase(unit.getLocation());
   if(distance <= attackDistance && distance > maxDistance) {
      returnedUnit = unit;
      maxDistance = distance;
   }
}
```

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,95 @@
+      Location pathLoc = new Location((int)(loc.getx()),
+                                      (int)(loc.gety())); 
+      int unitIndex = path.indexOf(pathLoc);
+      return Math.abs(index - unitIndex) - (loc.getDistance(pathLoc);
```
You don't need to make use `Math.abs` it should always be positive

thewells1024 replied:
you don't need to have the ( before `loc.getDistance(pathLoc)`

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,95 @@
+   private double distanceToBase(Location loc){
+      int index = path.size() - 1;
+      Location pathLoc = new Location((int)(loc.getx()),
+                                      (int)(loc.gety())); 
```
this shouldn't be indented that far, only 3 spaces

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,95 @@
+      int maxHealth = 0;
+      for(Unit unit : units){
+         double distance = unit.getLocation().getDistance(this.location);
+         int curHealth = unit.getHealth()
```
you need a semi-colon

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,97 @@
+   public List<Unit> selectFirstTarget(final List<Unit> units){
+      Unit returnedUnit;
+      double maxDistance = 0;
+      for(Unit unit: units) {
```
right now, this will return the last unit in range, not the first

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,97 @@
+
+   public List<Unit> selectLastTarget(final List<Unit> units){
+      Unit returnedUnit;
+      double minDistance = Integer.MAX_VALUE;
```
right now this will return the first unit in range

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,97 @@
+import java.util.List;
+import java.util.Collections;
+import java.util.ArrayList;
+
```
You need to import a lot of things, I would use Android Studio that will catch most of the issues you've been having

CVaranese replied:
I don't see a way to get info on this from android studio

thewells1024 replied:
Android Studio will show you when you're using a type you haven't imported and will try to suggest types to import

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,97 @@
+
+   public List<Unit> selectLastTarget(final List<Unit> units){
+      Unit returnedUnit;
+      double maxDistance = 0;
```
This should be `Integer.MIN_VALUE`

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,97 @@
+         double distance = distanceToBase(unit.getLocation());
+         double towerDistance = unit.getLocation().getDistance(this.location);
+         if(towerDistance <= attackDistance 
+            && distanceToBase(unit.getLocation()) > maxDistance){
```
don't recompute this, just use `distance > maxDistance`

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,97 @@
+         double distance = distanceToBase(unit.getLocation());
+         double towerDistance = unit.getLocation().getDistance(this.location);
+         if(towerDistance <= attackDistance 
+            && distanceToBase(unit.getLocation()) < minDistance){
```
don't recompute this, just use `distance > maxDistance`

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,104 @@
+import java.util.Collections;
+import java.util.ArrayList;
+
+import game.Location;
```
I believe this has to be fully qualified ie (cpe307.team6.toweroffense.game.Location)

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,104 @@
+@AllArgsConstructor
+public class SingleTargetTower implements Tower {
+   @Getter private final Location location;
+   private final static int attackDistance;
```
final static variables should be SCREAMING_SNAKE_CASE, and need to be given a value

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,104 @@
+public class SingleTargetTower implements Tower {
+   @Getter private final Location location;
+   private final static int attackDistance;
+   private final static int damage;
```
final static variables should be SCREAMING_SNAKE_CASE, and need to be given a value

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,104 @@
+   @Getter private final Location location;
+   private final static int attackDistance;
+   private final static int damage;
+   private final static Priority priority;
```
this shouldn't be final or static

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,104 @@
+   private final static int attackDistance;
+   private final static int damage;
+   private final static Priority priority;
+   private final static List<Location> path;
```
this shouldn't be static

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,104 @@
+
+@AllArgsConstructor
+public class SingleTargetTower implements Tower {
+   @Getter private final Location location;
```
this should be under the final static variables.

In general, variables should be final static, static, final, non-final

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,106 @@
+   private final static int ATTACK_DISTANCE = 0;
+   private final static int DAMAGE = 0;
+
+   private final List<Location> path = null;
```
don't initialize this, otherwise the path will always be null

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,106 @@
+
+@AllArgsConstructor
+public class SingleTargetTower implements Tower {
+   private final static int ATTACK_DISTANCE = 0;
```
just set this to an arbitrary non-zero value for now

thewells1024 commented on app/src/main/java/cpe307/team6/toweroffense/SingleTargetTower.java:
```diff
@@ -0,0 +1,106 @@
+@AllArgsConstructor
+public class SingleTargetTower implements Tower {
+   private final static int ATTACK_DISTANCE = 0;
+   private final static int DAMAGE = 0;
```
just set this to an arbitrary non-zero value for now
